Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -17,12 +17,86 @@
 #include "Firestore/core/src/firebase/firestore/remote/serializer.h"
 
 #include <gtest/gtest.h>
+#include <pb.h>
 #include <pb_encode.h>
 
+#include "Firestore/core/src/firebase/firestore/model/database_id.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+
+using firebase::firestore::model::DatabaseId;
+using firebase::firestore::model::FieldValue;
+using firebase::firestore::remote::Serializer;
+
 TEST(Serializer, CanLinkToNanopb) {
   // This test doesn't actually do anything interesting as far as actually using
   // nanopb is concerned but that it can run at all is proof that all the
   // libraries required for nanopb to work are actually linked correctly into
   // the test.
   pb_ostream_from_buffer(NULL, 0);
 }
+
+// Fixture for running serializer tests.
+class SerializerTest : public ::testing::Test {
+ public:
+  SerializerTest() : serializer(DatabaseId("p", "d")) {
+  }
+  Serializer serializer;
+
+  void ExpectRoundTrip(const FieldValue& model,
+                       const Serializer::ValueWithType& proto,
+                       FieldValue::Type type) {
+    EXPECT_EQ(type, model.type());
+    EXPECT_EQ(type, proto.type);
+    Serializer::ValueWithType actual_proto = serializer.EncodeFieldValue(model);
+    EXPECT_EQ(type, actual_proto.type);
+    EXPECT_EQ(proto, actual_proto);
+    EXPECT_EQ(model, serializer.DecodeFieldValue(proto));
+  }
+
+  void ExpectRoundTrip(const Serializer::ValueWithType& proto,
+                       const uint8_t* bytes,
+                       size_t bytes_len,
+                       FieldValue::Type type) {
+    EXPECT_EQ(type, proto.type);
+    // TODO(rsgowman): How big should this buffer be? Unclear; see TODO in
+    // remote/serializer.h on the Serializer::EncodeValueWithType() method.
+    // Hardcode to 1k for now.
+    uint8_t actual_bytes[1024];
```

Comment:
Can you also add a check to make sure bytes_len < 1024 to avoid overflow the buffer early (and potentially catch that explicitly, if that do happen).

Reply:
This is exactly the kind of thing I think we should eliminate by accumulating our results in a dynamic buffer. There's no reason to take on the complexity required to maintain these sizes across callers.